ENH pass env var to child

tomMoral opened the pull request:
Pass environment variable to child process first to avoid mixed effect after module loads.

TODO:

- [x] write tests
- [x] implement support for windows with the loky start method
- [x] make it possible to override the env of the children without touching the `os.environ` of the parent
- [x] do a manual check when calling a bunch of `np.linalg.svd` in a loky `executor.map` loop and check that there is no over-subscription on a machine with more than 32 cores.
- [ ] (optional) implement support for windows with the spawn start method.

tomMoral commented:
This seems to work well locally with the following script.

```python
import os
import numpy as np
from loky import get_reusable_executor

A = np.random.rand(10000 * 1000).reshape(10000, 1000)
r = np.linalg.svd(A)  # Uses 8 threads to compute the SVD

os.environ.update(OMP_NUM_THREADS='1', MKL_NUM_THREADS='1')
executor = get_reusable_executor(max_workers=4, reuse=False)
f = executor.map(np.linalg.svd, [A] * 4)  # Uses only 1 thread per process
r = list(f)
```

Capture of `htop` during the parallel computations:
![image](https://user-images.githubusercontent.com/3321081/64670601-03fa5180-d466-11e9-9c5d-8abca58249d3.png)

Without setting the environment variable, I can clearly see a large over-subscription.

tomMoral commented:
The decrease coverage come from lower number of lines and a cosmetic change.

ogrisel commented on loky/backend/fork_exec.py:
```diff
@@ -35,9 +35,12 @@ def close_fds(keep_fds):  # pragma: no cover
 def fork_exec(cmd, keep_fds):
 
+    # copy the environment variables to set in the child process
+    child_env = os.environ.copy()
```
What about making it possible to pass environment variables to override some variables only in the children without changing the `os.environ` dict of the main process?

This new `env` dict could be passed as an argument to the `get_reusable_executor` function.

tomMoral replied:
That was what I did before but I am rather hesitant as it means changing the API for `Process` compared to the one from `multiprocessing`. As we do not implement this change for other `start_method`, it would mean having a different way of starting process depending on the context.

ogrisel commented:
BTW, I have tested on a machine with 36 physical cores and it seems to work as expected (with htop).

ogrisel commented:
Interesting failure:

```
E       AssertionError: Expected var=loky rocks but got loky works
E       assert 'loky works' == 'loky rocks'
```

:)

ogrisel commented:
BTW, you don't need to put a msg that displays the content of the variables when doing `assert a == b` with pytest. pytest already reports the content of a and b.

ogrisel commented on tests/test_loky_backend.py:
```diff
@@ -279,6 +279,42 @@ def test_connection_queue(self):
+
+        # Test that the environment variable is correctly copied in the child
+        # process.
+        os.environ[key] = value
```
You should probably wrap this with a try / finally block to remove the `loky_child_env_process` environment variable after the end of the test to avoid any test side effect on other tests.

tomMoral replied:
done

tomMoral commented:
> Interesting failure:
> 
> ```
> E       AssertionError: Expected var=loky rocks but got loky works
> E       assert 'loky works' == 'loky rocks'
> ```
> 
> :)

The link to the [failing CI build](https://ci.appveyor.com/project/tomMoral/loky/builds/27479295/job/38gddxfr2n5dfv97). This is very weird as it seems to be random... I don't get what went wrong.

ogrisel commented on loky/backend/popen_loky_win32.py:
```diff
@@ -64,14 +65,15 @@ def __init__(self, process_obj):
 
+        # copy the environment variables to set in the child process
+        child_env = os.environ.copy()
+        child_env.update(process_obj.env)
```
Is the content of `process_obj.env` correct when the assertion fails under windows?

tomMoral replied:
I updated the test to print the content of `p.env` on failure and report on `xfail` so we should see if it is correct.

tomMoral replied:
Amazing report ...

```
========================== short test summary info ===========================
XFAIL tests/test_loky_backend.py::TestLokyBackend::test_child_env_process
  Can randomly fail on python2.7 and windows.
```

ogrisel commented:
If this test is only failing under Python 2.7 under Windows, maybe we should not waste time finding a workaround for a bug in `_winapi.CreateProcess` in Python 2.7 and just mark the test as xfail under Python 2.7 + windows.

ogrisel commented:
Also, switching to azure pipelines might speed-up the windows CI :)

tomMoral commented:
> If this test is only failing under Python 2.7 under Windows, maybe we should not waste time finding a workaround for a bug in `_winapi.CreateProcess` in Python 2.7 and just mark the test as xfail under Python 2.7 + windows.

It happened again [here](https://ci.appveyor.com/project/tomMoral/loky/builds/27488466/job/lscxfp71ajvjcbtg) in python2.7. I will mark the test with a `xfail` for this case.


![its happening again](http://www.davidlynch.it/wp-content/uploads/2016/07/Happening-again.gif)

ogrisel commented on CHANGES.md:
```diff
@@ -3,6 +3,9 @@
   a virtual environment (#216).
 
+- Copy the environment variables in the child process for ``LokyProcess``.
+  This allow setting env variables before loading any module. (#217)
```
This should be updated to reflect the fact that it's now also possible to pass specific environment variables just for the worker processes without changing the environment of the parent process.

ogrisel replied:
Maybe also add a disclaimer that this feature is does not work with Python 2.7 and Windows.

ogrisel commented on loky/process_executor.py:
```diff
@@ -853,7 +853,7 @@ class ProcessPoolExecutor(_base.Executor):
     def __init__(self, max_workers=None, job_reducers=None,
                  result_reducers=None, timeout=None, context=None,
-                 initializer=None, initargs=()):
+                 initializer=None, initargs=(), env={}):
```
Please do not use mutable datastructures as the default value for kwargs. This can lead to very subtle bugs.

Instead pass `env=None` and use a construct such as:

```python
env = {} if env is None else env
```

in the body of the method/function.

ogrisel commented on loky/process_executor.py:
```diff
@@ -874,6 +874,9 @@ def __init__(self, max_workers=None, job_reducers=None,
             initargs: A tuple of arguments to pass to the initializer.
+            env: A dict of environment variable to overwrite in the child
+                process. This only works with the loky context. The environment
+                variables are set before any module load.
```
Maybe add a comment stating that this does not work reliably under Python 2.7 with Windows.

ogrisel commented on tests/_test_process_executor.py:
```diff
@@ -892,3 +892,26 @@ def test_exited_child(self):
+
+        msg = "Expected env variable {}={}, got {}".format(var_name, var_value,
+                                                           var_child)
+        assert var_child == var_value, msg
```
As said elsewhere, I think you can just use:

```python
assert var_child == var_value
```

and drop the `msg`. pytest is clever enough to report informative error message.

ogrisel commented:
There was an xfail in the latest appveyor with Python 2.7 but it's passed otherwise:

https://ci.appveyor.com/project/tomMoral/loky/builds/27489597/job/s5xaapmlkb4qwh4i

tomMoral commented:
Ok, I updated the `CHANGE.md` and docstrings to correctly describe the feature. I also changed the appveyor CI to try to report the failure, and see if we are not missing something.

ogrisel commented on loky/reusable_executor.py:
```diff
@@ -73,6 +73,13 @@ def get_reusable_executor(max_workers=None, context=None, timeout=10,
+    ``VAR`` are string literals to overwrite the environment variable ``ENV``
+    in the child processes to value ``VAL``. The environment variables are set
+    in the children before any module is loaded. This only works with with the
+    ``loky`` context and it is unstable on windows with Python2.7.
```
```suggestion
    ``loky`` context and it is unreliable on Windows with Python 2.7.
```

ogrisel commented:
Everything was green, I merged without waiting CI for the last nitpick commit :)

Thank you very much @tomMoral. Shall we release loky?

tomMoral commented:
Yes I guess we should!
